sim/commands.go

```diff
@@ -76,31 +76,83 @@ func DoDisconnect(src, dst *Puppet) error {
 }
 
 func queryLatest(p *Puppet) ([]Latest, error) {
-	var empty interface{}
-	c, src, err := sourceRequest(p, muxrpc.Method{"replicate", "upto"}, empty)
+	// this is less efficient than using `replicate.upto`. replicate.upto however doesn't work in a ssb-db2 context, and
+	// nobody seems to want to patch that into ssb-db2 as a compat muxrpc so here we go
+
+	// gather all log stream responses we can find in a map of pubkey->largest seqno
+	seqnos := make(map[string]Latest)
+	type logStreamResponseValue struct {
+		Author    string
+		Sequence  int `json:"sequence"`
+		Timestamp int `json:"timestamp"`
+	}
+
+	type logStreamResponse struct {
+		Value logStreamResponseValue `json:"value"`
+	}
+
+	type sourceOptions struct {
+		Reverse bool `json:"reverse"`
+	}
+
+	opts := sourceOptions{Reverse: true}
+	c, src, err := sourceRequest(p, muxrpc.Method{"createLogStream"}, opts)
 	if err != nil {
-		return nil, err
+		return []Latest{}, err
 	}
 	defer c.Terminate()
 
-	var seqnos []Latest
 	ctx := context.TODO()
-	for src.Next(ctx) {
-		var l Latest
-		err = src.Reader(func(rd io.Reader) error {
-			return json.NewDecoder(rd).Decode(&l)
-		})
+
+	parseLogStream := func(rd io.Reader) error {
+		// read all the json-encoded data as bytes
+		b, err := io.ReadAll(rd)
 		if err != nil {
-			return nil, err
+			return err
+		}
+
+		// decode the response bits we care about
+		var resp logStreamResponse
+		err = json.Unmarshal(b, &resp)
+		if err != nil {
+			return err
 		}
-		seqnos = append(seqnos, l)
+
+		// field msg.value was omitted in log stream response => likely a response from the go stack; try to unpack as if
+		// the message response was `msg.value` itself
+		if resp.Value.Sequence == 0 && resp.Value.Author == "" {
+			var respGo logStreamResponseValue
+			err = json.Unmarshal(b, &respGo)
+			if err != nil {
+				return err
+			}
+			current, exist := seqnos[respGo.Author]
+			if !exist || respGo.Sequence > current.Sequence {
+				seqnos[respGo.Author] = Latest{ID: respGo.Author, Sequence: respGo.Sequence, TS: respGo.Timestamp}
+			}
+		} else { // unpack the js stack's response (which has the key msg.value)
+			// only update the map if the encountered sequence number is larger than what we already have stored
+			current, exist := seqnos[resp.Value.Author]
+			if !exist || resp.Value.Sequence > current.Sequence {
+				seqnos[resp.Value.Author] = Latest{ID: resp.Value.Author, Sequence: resp.Value.Sequence, TS: resp.Value.Timestamp}
+			}
```

Comment:
instead of doing this kludge i thought about implementing an interface on both `logStreamResponse` && `logStreamResponseValue`, triplet of getters for their values. felt overkill though, so i went with the approach above :~

Reply:
turns out this is really slow when dealing with a total of 150k messages ^^'